### ISSUE TITLE
Deobfuscate stack traces / crash reports

### ISSUE THREAD

**Issue by Marcono1234** [mojira/arisa-kt#606]
## The Problem
The fact that classes in the Minecraft JAR file are obfuscated has the following disadvantages:
- When a reporter just attaches a crash report without any additional information (possibly because they don't know when / why the game crashed), it is difficult to reason about what the cause might be
- Due to obfuscation mappings and line numbers changing between versions it can be difficult to tell whether two crash reports for different versions are caused by the same bug

It would therefore be useful if Arisa could deobfuscate stack traces / crash reports using the mapping files provided by Mojang.

## Possible Implementation
Initially it might be easiest to apply deobfuscation only to attached crash report files; detecting crash reports (or stack traces) in the description and comments might be difficult (and also Minecraft version information might be missing).

For deobfuscation ProGuard's [ReTrace](https://github.com/Guardsquare/proguard/blob/master/retrace/src/proguard/retrace/ReTrace.java) could be used.

Then, when either a new attachment is attached, or a report with attachments is created:
1. Check if the attachment is a crash report
2. Try to extract version information from the crash report (relying on Affected Versions of the issue might be error-prone)
3. Verify that the version has obfuscation mappings; i.e. version >= 1.14.4 or 19w36a
4. Check if the mapping file is already present on the file system
    1. If not, download the mapping file (perform sanity checks to prevent writing the mapping file outside of intented directory)
5. Deobfuscate the crash report
6. Attach the crash report as <code>(deobfuscated) <i>original-crash-report-name.txt</i></code>
(Maybe perform sanity check to prevent overwriting or shadowing existing attachment with that name)

## Considerations
- Verify that this use case is allowed by the license, in doubt get official confirmation from Mojang developers
- The obfuscation mapping license changed a few times (see also [Minecraft Wiki](https://minecraft.fandom.com/wiki/Obfuscation_map#License)), maybe deobfuscation for certain versions needs to be disabled if the license does not permit deobfuscation
- Check whether ReTrace can directly be applied to a crash report, or whether stack traces have to be extracted manually before being passed to ReTrace
- Check whether ["verbose" mode](https://github.com/Guardsquare/proguard/blob/master/docs/md/manual/tools/retrace.md#restoring-a-stack-trace-with-line-numbers-verbose--withverbose) of ReTrace might be useful (mainly for inspecting respective decompiled Minecraft source code more easily)
- ReTrace detects obfuscated class names using regex, this might lead to false positives when applying it to a complete crash report (since it contains more than just stack traces); though maybe this is negligible
- Maybe integrate this as part of https://github.com/urielsalis/mc-crash-lib ? Though arisa-kt would still be responsible for storing and providing mapping files then, and since most of the work is probably performed by ReTrace it might be easiest to have all code in arisa-kt only

**Comment by urielsalis** [comment]
I would do this in mc-crash-lib, as that's the library we are using to detect crashes already

**Comment by Marcono1234** [comment]
My main point was that mapping files probably should be saved to the file system (because ReTrace apparently only support a `File`, but not a stream or reader; and to avoid downloading them for every crash report), and how obfuscation files are stored or where they come from should probably not be the responsibility of mc-crash-lib.

But then, the task that mc-crash-lib would need to perform would be rather small (simplified):
1. arisa-kt: `AttachmentUtils.extractCrashesFromAttachments`
2. arisa-kt: Filter for `Crash.Minecraft`, check version, download obfuscation mapping if necessary
3. mc-crash-lib: Deobfuscate
4. arisa-kt: Attach deobfuscated crash report

I am not sure if it would really be worth it adding this single task to mc-crash-lib, or if it would be easier to just have this in arisa-kt. Step 1 (partially) and step 2 could possibly also be performed by mc-crash-lib, but then we would have duplicate code (unless we decide to move some of `AttachmentUtils`'s code to mc-crash-lib).

(Either way it might be good to leave this issue here open since some implementation has to be performed in arisa-kt.)

**Comment by violine1101** [comment]
> The obfuscation mapping license changed a few times (see also Minecraft Wiki), maybe deobfuscation for certain versions needs to be disabled if the license does not permit deobfuscation

This is not an issue. All versions that have mappings currently have the same license; it seems like the new license was applied to the old mappings retroactively. The wiki appears to be wrong about this. ([Here's](https://minecraft.fandom.com/wiki/Minecraft_Wiki:Java_Edition_version_list) a collection of all the mapping links, however this only works when logged into the wiki)

What is an issue is that not all versions have a map, but that should not be difficult to consider.

IMO this should completely be done in mc-crash-lib, without Arisa doing anything with the crash report directly, that way we keep that complexity out of here.

**Comment by urielsalis** [comment]
Agreed with violine, mc-crash-lib is also going to be used in other bots so
its nice to have it on there

**Comment by Marcono1234** [comment]
Thanks, looks good!
